### PR TITLE
feat: Refactor kubectl-crane plugin

### LINKED ISSUE #592: Proposal: Refactor kubectl-crane plugin
## Describe the feature

In order to implement the goal #523 described, I create this proposal to refactor [kubectl-crane](https://github.com/gocrane/kubectl-crane) plugin, and describe it as below:

1. `kubectl-crane rr list`
Get recommended rules
Supported parameters：
- `--name` specify name for `recommendationrules` 
- `--recommender` specify type for `recommendationrules` 

2. `kubectl-crane rr create`
Create a template for `recommendationrules`.
Supported parameters：
- `--recommender` specify type for `recommendationrules`，separated with `,` if more than one 。default is `Resource`
- `--target` specify recommend target
- `--namespace` specify namespace, default is `Any`
- `runInterval` specify `runInterval `

3. `kubectl-crane recommend list {type}`

`type` recommend type[Resource, Replicas, IdleNode]
Get recommend list.
Supported parameters:
- `--name` specify name for `recommend`
- `--namespace` specify namespace for `recommend`
- `--targetKind` specify target type for `recommend`

5. `kubectl-crane recommend adopt {name}`
Adopt a recommendation.
Supported parameters:
- `--dryRun` pre-commit

6. `kubectl-crane view-recommend`
View a source which recommends related.
Supported parameters:
- `--name` specify resource name
- `--namespace` specify resource namespace
- `--apiVersion` specify resource apiVersion
OR
- `--selector` like `'{"apiVersion":"","kind": "", "name": "", "namespace":""}'`

Please join us if you have other suggestion or interested this proposal.

### PR DESCRIPTION
**Pull request by mangoGoForward**
Signed-off-by: mango <xu.weiKyrie@foxmail.com>

Fixes gocrane/crane#592 gocrane/crane#523

### CODE REVIEW

**Comment by mangoGoForward** [comment]
@qmhu Please have a review when you have free time

**Comment by qmhu** [comment]
Thanks for you effort, Looks neat! @mangoGoForward

File: pkg/cmd/crane.go [hunk 1]
```diff
@@ -18,22 +18,21 @@ package cmd
 
 import (
 	"fmt"
+	"github.com/gocrane/kubectl-crane/pkg/cmd/options"
```
Comment (qmhu) [inline pkg/cmd/crane.go]: go imports should follow the guide: https://gocrane.io/docs/contributing/code-standards/

File: pkg/cmd/crane.go [hunk 2]
```diff
@@ -18,22 +18,21 @@ package cmd
...
-	_ "k8s.io/client-go/plugin/pkg/client/auth"
-	"k8s.io/client-go/rest"
-
 	"k8s.io/cli-runtime/pkg/genericclioptions"
+	_ "k8s.io/client-go/plugin/pkg/client/auth"
 )
 
 var (
 	craneExample = `
-	# view view pod resource recommendation
```
Comment (qmhu) [inline pkg/cmd/crane.go:32]: remove unused content  like `craneExample `

File: pkg/cmd/recommend/adopt.go [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+}
+
+func (o *RecommendAdoptOptions) Run() error {
+	recommend, err := o.CommonOptions.CraneClient.AnalysisV1alpha1().Recommendations("").Get(context.TODO(), o.Name, metav1.GetOptions{})
+	if err != nil {
+		return errors.New("the recommend doesn't exist, please specify a existed recommend name with --name")
+	}
+
+	patchOptions := metav1.PatchOptions{}
+	if len(o.DryRun) != 0 {
```
Comment (qmhu) [inline pkg/cmd/recommend/adopt.go]: Can we print the output for dryrun?

File: pkg/cmd/recommend/adopt.go [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+		patchOptions.DryRun = []string{"All"}
+	}
+
+	if recommend.Spec.TargetRef.Kind == "Deployment" {
```
Comment (qmhu) [inline pkg/cmd/recommend/adopt.go]: Please refer to https://github.com/gocrane/crane/blob/main/pkg/server/handler/recommendation/recommendation.go#L117
to implement how to adopt a recommendation.

File: pkg/cmd/recommend/adopt.go [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+			return errors.New(fmt.Sprintf("adopt the recommend failed because %v", err))
+		}
+	}
+
+	return nil
+}
+
+func (o *RecommendAdoptOptions) AddFlags(cmd *cobra.Command) {
+	cmd.Flags().StringVarP(&o.Name, "name", "", "", "Specify the name for recommend")
+	cmd.Flags().StringVarP(&o.DryRun, "dryRun", "", "", "Pre-commit")
```
Comment (qmhu) [inline pkg/cmd/recommend/adopt.go]: all flags should be use 短划线 like "dry-run"

File: pkg/cmd/recommend/list.go [hunk 6]
```diff
@@ -0,0 +1,219 @@
...
+func (o *RecommendListOptions) Complete(cmd *cobra.Command, args []string) error {
+	if err := o.CommonOptions.Complete(cmd, args); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (o *RecommendListOptions) Run() error {
+	query := util.NewQuery()
```
Comment (qmhu) [inline pkg/cmd/recommend/list.go]: No need to use a custom query

File: pkg/cmd/recommend/list.go [hunk 7]
```diff
@@ -0,0 +1,219 @@
...
+		query.Filters[util.FieldName] = util.Value(o.Name)
+	}
+
+	namespace := ""
+	if len(*o.CommonOptions.ConfigFlags.Namespace) > 0 {
+		namespace = *o.CommonOptions.ConfigFlags.Namespace
+		//query.Filters[util.FieldNamespace] = util.Value(*o.CommonOptions.ConfigFlags.Namespace)
+	}
+
+	recommendResult, err := o.CommonOptions.CraneClient.AnalysisV1alpha1().Recommendations(namespace).List(context.TODO(), metav1.ListOptions{})
```
Comment (qmhu) [inline pkg/cmd/recommend/list.go]: Recommendation have enough labels to support label filter. Using label filter is faster and effective. 
Also we can support more flags to query recommendation:
- RecommendationRule 名称：analysis.crane.io/recommendation-rule-name
- RecommendationRule UID：analysis.crane.io/recommendation-rule-uid
- RecommendationRule 的 recommender：analysis.crane.io/recommendation-rule-recommender
- 推荐资源的 kind：analysis.crane.io/recommendation-target-kind
- 推荐资源的 apiversion：analysis.crane.io/recommendation-target-apiversion
- 推荐资源的 name：analysis.crane.io/recommendation-target-apiversion

File: pkg/cmd/recommend/list.go [hunk 8]
```diff
@@ -0,0 +1,219 @@
...
+
+	for _, recommendation := range recommendations {
+		row := table.Row{}
+
+		row = append(row, recommendation.Name)
+		row = append(row, recommendation.Spec.TargetRef.Name)
+		row = append(row, recommendation.Namespace)
+		row = append(row, recommendation.Spec.TargetRef.Kind)
+
+		currentInfo := v1.Deployment{}
```
Comment (qmhu) [inline pkg/cmd/recommend/list.go]: Suggest to copy structs in craned and move to apis:
https://github.com/gocrane/crane/blob/main/pkg/recommendation/recommender/resource/recommend.go#L24
https://github.com/gocrane/crane/blob/main/pkg/recommendation/recommender/replicas/recommend.go#L22

File: pkg/cmd/recommend/list.go [hunk 9]
```diff
@@ -0,0 +1,219 @@
...
+					currentResource += container.Name + "/" + container.Resources.Requests.Cpu().String() + "m/" + container.Resources.Requests.Memory().String() + "\n"
+				}
+			} else if recommendation.Spec.Type == "Replicas" {
+				currentResource += strconv.Itoa(int(*currentInfo.Spec.Replicas))
+			}
+
+			row = append(row, currentResource)
+		}
+
+		recommendInfo := v1.Deployment{}
```
Comment (qmhu) [inline pkg/cmd/recommend/list.go]: No need to know the actual workload types.

File: pkg/cmd/recommendationRule/create.go [hunk 10]
```diff
@@ -0,0 +1,161 @@
...
+	}
+
+	klog.Infof(fmt.Sprintf("the recommendation rule %s created successfully", name))
+	return nil
+}
+
+func (o *RecommendationRuleCreateOptions) AddFlags(cmd *cobra.Command) {
+	cmd.Flags().StringVarP(&o.Recommender, "recommender", "", "Resource", "specify type for recommendationrules，separated with ',' if more than one, default is Resource")
+	cmd.Flags().StringVarP(&o.Target, "target", "", "", "specify recommend target for recommendationrules")
+	cmd.Flags().StringVarP(&o.RunInterval, "runInterval", "", "", "Specify runInterval for recommendationrules")
```
Comment (qmhu) [inline pkg/cmd/recommendationRule/create.go]: Use `run-interval` and `dry-run`

File: pkg/cmd/recommendationRule/create.go [hunk 11]
```diff
@@ -0,0 +1,161 @@
...
+
+	recommendationRule.Spec.RunInterval = o.RunInterval
+	name += "-" + o.RunInterval
+	name += "-" + time.Now().Format("2006-01-02")
+
+	recommendationRule.Name = name
+
+	createOptions := metav1.CreateOptions{}
+	if len(o.DryRun) != 0 {
+		createOptions.DryRun = []string{"All"}
```
Comment (qmhu) [inline pkg/cmd/recommendationRule/create.go:155]: If dryRun, then print the object result like native kubectl

File: pkg/cmd/view_recommend.go [hunk 12]
```diff
@@ -0,0 +1,188 @@
...
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/klog/v2"
+	"reflect"
+	"strconv"
+)
+
+var (
+	viewRecommendExample = `
+# view all recommend result with kube-system namespace
+%[1]s view-recommend --sourceSelector '{"apiVersion":"","kind": "", "name": "", "namespace":""}'
```
Comment (qmhu) [inline pkg/cmd/view_recommend.go]: how about : [1]s view-recommend  --api-version v1 --kind Deployment -n {namespace} {name}

File: pkg/cmd/view_recommend.go [hunk 13]
```diff
@@ -0,0 +1,188 @@
...
+
+	return nil
+}
+
+func (o *ViewRecommendOptions) renderTable(recommendations []analysisv1alph1.Recommendation) {
+	t := table.NewWriter()
+	t.SetStyle(table.StyleLight)
+	t.SetOutputMirror(o.CommonOptions.Out)
+	header := table.Row{}
+	header = append(header, table.Row{"NAME", "RECOMMEND SOURCE", "NAMESPACE", "TARGET", "CURRENT RESOURCE", "RECOMMEND RESOURCE", "CREATED TIME", "UPDATED TIME"}...)
```
Comment (qmhu) [inline pkg/cmd/view_recommend.go]: Better to reuse the code for all table render. for example, recommendation table should use the same code , the same as recommendationrule table.

File: pkg/cmd/crane.go (hunk 1 above)
Comment (mangoGoForward) [reply pkg/cmd/crane.go]: Done

File: pkg/cmd/crane.go (hunk 2 above)
Comment (mangoGoForward) [reply pkg/cmd/crane.go:32]: Done

File: pkg/cmd/recommend/adopt.go (hunk 4 above)
Comment (mangoGoForward) [reply pkg/cmd/recommend/adopt.go]: Done

File: pkg/cmd/recommend/adopt.go (hunk 5 above)
Comment (mangoGoForward) [reply pkg/cmd/recommend/adopt.go]: Done

File: pkg/cmd/recommendationRule/create.go (hunk 10 above)
Comment (mangoGoForward) [reply pkg/cmd/recommendationRule/create.go]: Done

File: pkg/cmd/view_recommend.go (hunk 13 above)
Comment (mangoGoForward) [reply pkg/cmd/view_recommend.go]: Done

File: pkg/cmd/recommend/list.go (hunk 7 above)
Comment (mangoGoForward) [reply pkg/cmd/recommend/list.go]: Done

File: pkg/cmd/recommend/list.go (hunk 9 above)
Comment (mangoGoForward) [reply pkg/cmd/recommend/list.go]: Done

**Review by qmhu — APPROVED** [review]

**Pull request merged**